### DISCUSSION TITLE
State Machine!

### DISCUSSION THREAD

**Discussion by lukasvrenner** [lukasvrenner/turtls#23] [Ideas]
I'm currently in the process of rewriting the library structure to store all state within the `Connection` struct. IO operations are fallible, and it makes sense for the application to decide how to proceed. `turtls_connect` will return a special error on IO failure. If the application decides the error is non-fatal, they can call `turtls_connect` again. It will resume the handshake where it left off. Additionally, once `turtls_read` and `turtls_write` (exact names TBD) are implemented, they can automatically perform the handshake if it hasn't already occurred. This is how OpenSSL does it, and I think it makes a lot of sense.

**Comment by lukasvrenner** [comment]
The new implementation is now at the same stage of functionality as before the state machine rewrite. ClientHello is sent and ServerHello is parsed. Handshake traffic secrets are calculated, and the EncryptedExtensions are successfully decrypted. Now, if an IO error occurs, the application can decide to resume the handshake.

This isn't quite true yet. None of the errors are properly handled and cleanup does not happen. Theoretically any error is resumable currently, which is a problem for obvious reasons.

# There goes readability
The source is much less readable. The issue is that control flow is no longer sequential. Rather than flowing straight down a file, each stage is now just a function that updates the state machine. After each stage, the state machine checks the current state and calls the next function accordingly. This can make control flow harder to follow.

This has an additional effect. A common pattern has emerged that results in egregious amounts of indentation:
```rust
impl Foo {
    fn bar(&mut self) {
        loop {
            match self.baz {
                Baz::Biz(biz) => {
                    if biz == x {
                        ...
                    }
                    ...
                },
                _ => ...,
            },
        }
    }
}
```
The only real solution I can think of to this problem is to extract the inner parts into separate functions. However this would further decrease readability by fragmenting functions even more.

**Reply by jjohnson-47** [reply]
Awesome progress here. This is outside of my expertise, but here's an AI response from o1, openai's newest model (with some prompt engineering) which I've found to be helpful with my own code issues. Infact, some of the suggestions below are helpful to me. I'm putting this in markdown on my phone, hopefully that renders here. 

# General Advice

This is common because a state machine makes control flow explicit rather than relying on linear, top-down progression. The keys to mitigating these issues are careful structuring, modular design, and thoughtful error handling strategies. (Easy to say:))

## Suggestions

### 1. Embrace Smaller, Focused Functions

Splitting large functions into many small functions can reduce readability, but it can actually improve it if done thoughtfully. Instead of one large, deeply nested function, consider having multiple well-named functions that each handle a single stage or a single decision. For example:

```rust
process_server_hello(&mut self, data: &ServerHello) -> Result<NextState, TlsError>

decrypt_encrypted_extensions(&mut self, data: &EncryptedExtensions) -> Result<NextState, TlsError>

calculate_handshake_secrets(&mut self) -> Result<NextState, TlsError>
```

Even if this fragments the code, the improved clarity of intent can outweigh the cost of having multiple files or functions. Each function's name and signature acts as documentation, making it easier to navigate and reason about the code.

### 2. Use the State Pattern or Enum Variants for States

A key goal is to reduce the complexity of large match blocks that span multiple indentation levels. Instead of a single large loop with a match self.baz, consider an approach where each state is either:

- Represented by its own struct implementing a State trait, or
- Encoded as an enum variant with associated data and a single function that transitions to the next state.

For example:

```rust
enum HandshakeState {
    WaitingForServerHello,
    ProcessingServerHello(ServerHelloData),
    DecryptingExtensions(ExtensionsData),
    // ...
}

impl HandshakeState {
    fn advance(self, context: &mut TlsContext) -> Result<Self, TlsError> {
        match self {
            HandshakeState::WaitingForServerHello => context.handle_waiting_for_server_hello(),
            HandshakeState::ProcessingServerHello(data) => context.handle_processing_server_hello(data),
            HandshakeState::DecryptingExtensions(data) => context.handle_decrypting_extensions(data),
        }
    }
}
```

Each method can be short and descriptive, reducing massive indentation and making control flow explicit through state transitions rather than deeply nested logic.

### 3. Clear Separation of Concerns

Organize the code so that each file or module corresponds to a logical phase of the handshake. For example:

- `client_hello.rs` handles everything related to generating and sending the ClientHello.
- `server_hello.rs` deals with receiving, parsing, and reacting to the ServerHello.
- `encrypted_extensions.rs` handles decryption and interpretation of EncryptedExtensions.

This way, readers know where to look for logic related to a particular handshake phase. Within each module, the logic can be broken into steps or functions for clarity.

### 4. Error Handling and Cleanup Strategies

Right now, any error might be considered "resumable." Instead of trying to handle all errors at the same point, consider designing your state transitions so that error conditions naturally lead to a "Failed" or "Cleanup" state. For example:

```rust
enum HandshakeState {
    WaitingForServerHello,
    ProcessingServerHello,
    Failed(TlsError),
    // ...
}
```

When an error occurs, transition the state machine to a Failed state. This state might run cleanup code or prepare a way to resume, if possible. This approach makes error conditions explicit in the state machine, rather than as ad-hoc exceptions in the code.

### 5. Commenting and High-Level Documentation

Because state-based code jumps around rather than flowing top-down, ensure you have ample high-level documentation. A "map" of the handshake process can help:

- Which states exist?
- What transitions can occur from one state to another?
- What errors or special conditions cause a fallback?

A short, well-maintained state diagram (even as a code comment) can work wonders for maintainability and readability. It helps developers quickly grasp the "big picture" before diving into individual functions.

### 6. Consider a State Machine Framework or Macros

Rust has crates like typestate or similar DSLs that can help structure state machines in a more declarative manner. These frameworks can help reduce boilerplate and keep your state transitions more readable. Even if you don't use a full library, careful use of traits, macros, or code generation might help reduce repetitive code.

### 7. Iterative Improvement

Don't expect perfect readability immediately after such a major refactor. State machines often start out looking fragmented but become clearer as you polish the design. Continue refining function names, documentation, and modular boundaries until the flow of the handshake states is intuitive.

**Reply by lukasvrenner** [reply]
Interestingly, this problem could be solved more elegantly in C via the `switch` statement. C's `switch` statement functions like a conditional `goto`. Because of the fallthrough behavior of `switch`, the entire loop can be eliminated. The state only needs to be checked when the function is called, but there is only one possible next step after the successful completion of any given step. It's possible the Rust compiler can recognize this and unroll the loop, but the Rust version is still syntactical much clunkier.

```c
switch (state->next_step) {
case CLIENT_HELLO:
    client_hello(state);
    /* falls through */
case SERVER_HELLO:
    server_hello(state);
    /* falls through */
    ...
}
```
This is arguably much easier to follow because it linearizes/defragments control flow and reduces indentation.
